src/server/scripts/Events/hallows_end.cpp

```diff
@@ -57,13 +78,391 @@ class spell_hallows_end_has_water_bucket : public SpellScriptLoader
+        {
+            switch (pos)
+            {
+                case 235431: x = -9445.1f; y = 63.27f; z = 58.16f; break;
+                case 235432: x = -5616.30f; y = -481.89f; z = 398.99f; break;
+                case 235433: x = -4198.1f; y = -12509.13f; z = 46.6f; break;
+                case 235434: x = 360.9f; y = -4735.5f; z = 11.773f; break;
+                case 235435: x = 2229.4f; y = 263.1f; z = 36.13f; break;
+                case 235436: x = 9532.9f; y = -6833.8f; z = 18.5f; break;
+            }
+        }
+
+        void Reset() override
+        {
+            events.ScheduleEvent(EVENT_TALK_1, Seconds(3));
+            events.ScheduleEvent(EVENT_APPLY_FIRE, Seconds(5));
+            events.ScheduleEvent(EVENT_STOP_APPLYING_FIRE, Minutes(2));
+            events.ScheduleEvent(EVENT_FINISH, Minutes(5));
+
+            FinishEvent(false);
+        }
+
+        void UpdateAI(uint32 diff) override
+        {
+            events.Update(diff);
+
+            if (me->HasUnitState(UNIT_STATE_CASTING))
+                return;
+
+            while (uint32 eventId = events.ExecuteEvent())
+            {
+                switch (eventId)
+                {
+                    case EVENT_TALK_1:
+                        Talk(0);
+                        break;
+                    case EVENT_APPLY_FIRE:
+                    {
+                        if (canShootFire)
+                        {
+                            if (Creature* trigger = me->FindNearestCreature(NPC_FIRE_TRIGGER, 50.0f))
+                                if (!trigger->GetAura(SPELL_START_FIRE) && !trigger->GetAura(SPELL_SPREAD_FIRE))
+                                    me->CastSpell(trigger, SPELL_START_FIRE);
+                            events.Repeat(Seconds(5));
+                        }
+                        break;
+                    }
+                    case EVENT_STOP_APPLYING_FIRE:
+                        canShootFire = false;
+                        break;
+                    case EVENT_FINISH:
+                    {
+                        bool failed = false;
+                        std::list<Creature*> triggerList;
+                        me->GetCreatureListWithEntryInGrid(triggerList, NPC_FIRE_TRIGGER);
+                        for (Creature* trigger : triggerList)
+                        {
+                            if (trigger->HasAura(SPELL_FIRE_AURA_BASE))
+                                failed = true;
+
+                            trigger->AI()->EnterEvadeMode();
+                        }
+                        me->StopMoving();
+                        FinishEvent(failed);
+                        break;
+                    }
+                    case EVENT_CLEAVE:
+                        me->CastSpell(me->GetVictim(), SPELL_CLEAVE);
+                        events.Repeat(Seconds(20));
+                        break;
+                }
+
+            }
+
+            if (!UpdateVictim())
```

Comment:
Why? Other scripts do the same when there are parts of the update that should only happen while in combat.